Experienced '504 Deadline Exceeded' error when tried to delete large number rows by partitioned dml

There are 1 billion old records need to be cleaned. According the documents, It seems Partitioned DML should be the good choice. I choose java client because we can set timeout by using setPartitionedDmlTimeout(Duration.ofHours(24L)). However, I still experienced the '504 Deadline Exceeded' after a few hours that shorter than 24 hours I set.  Why setPartitionedDmlTimeout did not work? How should I solve the problem? Is it not possible to delete so large number of data by using Partitioned DML. 

Following are the details.
#### Environment details

1. Spanner 
    3 nodes, 2 billion records in a table and 1 billion old records need to be cleaned
2. OS type and version: MacOs version 10.15.3
3. Java version: java 8
4. spanner version(s): the latest version 

#### Steps to reproduce

  1. Write 2 billion records with timestamp in to the spanner table with dataflow
  2. Run a clean job to delete  the old 1 billion records with Partitioned DML by java client
  3. Set the timeout to be 24 hours by using setPartitionedDmlTimeout(Duration.ofHours(24L))
  4. Run the java code

#### Code example

```java
SpannerOptions options = SpannerOptions.newBuilder().setPartitionedDmlTimeout(Duration.ofHours(24L)).build();

        Spanner spanner = options.getService();
        try {

            DatabaseId db = DatabaseId.of(options.getProjectId(), args[0], args[1]);
            // [END init_client]
            //This will return the default project id based on the environment.
            String clientProject = spanner.getOptions().getProjectId();
            if (!db.getInstanceId().getProject().equals(clientProject)) {
                System.err.println(
                        "Invalid project specified. Project in the database id should match the"
                                + "project name set in the environment variable GOOGLE_CLOUD_PROJECT. Expected: "
                                + clientProject);
            }

            // [START init_client]
            DatabaseClient dbClient = spanner.getDatabaseClient(db);

            SimpleDateFormat formatter = new SimpleDateFormat("dd/MM/yyyy hh:mm:ss.SSS");
            String sql="DELETE FROM MyTable WHERE updateTimestamp<'2020-05-02T03:00:00Z'";
            Date currentDate = new Date();
            System.out.println(String.format("Delete Start:%s",formatter.format(currentDate)));

            long lStartTime = Instant.now().toEpochMilli();
            deleteUsingPartitionedDml(dbClient,sql);
            long lEndTime = Instant.now().toEpochMilli();
            long elaspedTime = lEndTime - lStartTime;
            System.out.println("Elapsed time in : " + elaspedTime/1000);

        } finally {
            spanner.close();
        }
        // [END init_client]
        System.out.println("Closed client");
    }
    private static void deleteUsingPartitionedDml(DatabaseClient dbClient,String sql) {
        long rowCount = dbClient.executePartitionedUpdate(Statement.of(sql));

        System.out.printf("%d records deleted.\n", rowCount);
    }

```

@olavloite I tried to run the delete job again. I got the  'DEADLINE_EXCEEDED' instead of '504 DEADLINE_EXCEEDED'  about four hours later after started the delete job.

Following are the details.
1. Half of the data was deleted.
2. The stacktrace of the exception i got
`com.google.cloud.spanner.SpannerException: DEADLINE_EXCEEDED: com.google.api.gax.rpc.DeadlineExceededException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 3599999079514ns. [buffered_nanos=186035342, remote_addr=spanner.googleapis.com/172.217.175.106:443]
    at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerExceptionPreformatted (SpannerExceptionFactory.java:182)
    at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException (SpannerExceptionFactory.java:45)
    at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException (SpannerExceptionFactory.java:139)
    at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.get (GapicSpannerRpc.java:691)
    at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.executePartitionedDml (GapicSpannerRpc.java:543)
    at com.google.cloud.spanner.PartitionedDMLTransaction$1.call (PartitionedDMLTransaction.java:87)
    at com.google.cloud.spanner.PartitionedDMLTransaction$1.call (PartitionedDMLTransaction.java:69)
    at com.google.api.gax.retrying.DirectRetryingExecutor.submit (DirectRetryingExecutor.java:105)
    at com.google.cloud.RetryHelper.run (RetryHelper.java:76)
    at com.google.cloud.RetryHelper.runWithRetries (RetryHelper.java:50)
    at com.google.cloud.spanner.SpannerRetryHelper.runTxWithRetriesOnAborted (SpannerRetryHelper.java:47)
    at com.google.cloud.spanner.PartitionedDMLTransaction.executePartitionedUpdate (PartitionedDMLTransaction.java:91)
    at com.google.cloud.spanner.SessionImpl.executePartitionedUpdate (SessionImpl.java:103)
    at com.google.cloud.spanner.SessionPool$PooledSession.executePartitionedUpdate (SessionPool.java:647)
    at com.google.cloud.spanner.DatabaseClientImpl$3.apply (DatabaseClientImpl.java:199)
    at com.google.cloud.spanner.DatabaseClientImpl$3.apply (DatabaseClientImpl.java:196)
    at com.google.cloud.spanner.DatabaseClientImpl.runWithSessionRetry (DatabaseClientImpl.java:213)
    at com.google.cloud.spanner.DatabaseClientImpl.executePartitionedUpdate (DatabaseClientImpl.java:194)
    at jp.co.sej.central.spanner.DeleteOldStoreRecords.deleteUsingPartitionedDml (DeleteOldStoreRecords.java:93)
    at jp.co.sej.central.spanner.DeleteOldStoreRecords.main (DeleteOldStoreRecords.java:81)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
    at java.lang.Thread.run (Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: com.google.api.gax.rpc.DeadlineExceededException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 3599999079514ns. [buffered_nanos=186035342, remote_addr=spanner.googleapis.com/172.217.175.106:443]
    at com.google.common.util.concurrent.AbstractFuture.getDoneValue (AbstractFuture.java:552)
    at com.google.common.util.concurrent.AbstractFuture.get (AbstractFuture.java:533)
    at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.get (GapicSpannerRpc.java:685)
    at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.executePartitionedDml (GapicSpannerRpc.java:543)
    at com.google.cloud.spanner.PartitionedDMLTransaction$1.call (PartitionedDMLTransaction.java:87)
    at com.google.cloud.spanner.PartitionedDMLTransaction$1.call (PartitionedDMLTransaction.java:69)
    at com.google.api.gax.retrying.DirectRetryingExecutor.submit (DirectRetryingExecutor.java:105)
    at com.google.cloud.RetryHelper.run (RetryHelper.java:76)
    at com.google.cloud.RetryHelper.runWithRetries (RetryHelper.java:50)
    at com.google.cloud.spanner.SpannerRetryHelper.runTxWithRetriesOnAborted (SpannerRetryHelper.java:47)
    at com.google.cloud.spanner.PartitionedDMLTransaction.executePartitionedUpdate (PartitionedDMLTransaction.java:91)
    at com.google.cloud.spanner.SessionImpl.executePartitionedUpdate (SessionImpl.java:103)
    at com.google.cloud.spanner.SessionPool$PooledSession.executePartitionedUpdate (SessionPool.java:647)
    at com.google.cloud.spanner.DatabaseClientImpl$3.apply (DatabaseClientImpl.java:199)
    at com.google.cloud.spanner.DatabaseClientImpl$3.apply (DatabaseClientImpl.java:196)
    at com.google.cloud.spanner.DatabaseClientImpl.runWithSessionRetry (DatabaseClientImpl.java:213)
    at com.google.cloud.spanner.DatabaseClientImpl.executePartitionedUpdate (DatabaseClientImpl.java:194)
    at jp.co.sej.central.spanner.DeleteOldStoreRecords.deleteUsingPartitionedDml (DeleteOldStoreRecords.java:93)
    at jp.co.sej.central.spanner.DeleteOldStoreRecords.main (DeleteOldStoreRecords.java:81)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
    at java.lang.Thread.run (Thread.java:748)
Caused by: com.google.api.gax.rpc.DeadlineExceededException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 3599999079514ns. [buffered_nanos=186035342, remote_addr=spanner.googleapis.com/172.217.175.106:443]
    at com.google.api.gax.rpc.ApiExceptionFactory.createException (ApiExceptionFactory.java:51)
    at com.google.api.gax.grpc.GrpcApiExceptionFactory.create (GrpcApiExceptionFactory.java:72)
    at com.google.api.gax.grpc.GrpcApiExceptionFactory.create (GrpcApiExceptionFactory.java:60)
    at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure (GrpcExceptionCallable.java:97)
    at com.google.api.core.ApiFutures$1.onFailure (ApiFutures.java:68)
    at com.google.common.util.concurrent.Futures$CallbackListener.run (Futures.java:982)
    at com.google.common.util.concurrent.DirectExecutor.execute (DirectExecutor.java:30)
    at com.google.common.util.concurrent.AbstractFuture.executeListener (AbstractFuture.java:1138)
    at com.google.common.util.concurrent.AbstractFuture.complete (AbstractFuture.java:957)
    at com.google.common.util.concurrent.AbstractFuture.setException (AbstractFuture.java:748)
    at io.grpc.stub.ClientCalls$GrpcFuture.setException (ClientCalls.java:520)
    at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose (ClientCalls.java:495)
    at io.grpc.PartialForwardingClientCallListener.onClose (PartialForwardingClientCallListener.java:39)
    at io.grpc.ForwardingClientCallListener.onClose (ForwardingClientCallListener.java:23)
    at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose (ForwardingClientCallListener.java:40)
    at com.google.cloud.spanner.spi.v1.SpannerErrorInterceptor$1$1.onClose (SpannerErrorInterceptor.java:100)
    at io.grpc.PartialForwardingClientCallListener.onClose (PartialForwardingClientCallListener.java:39)
    at io.grpc.ForwardingClientCallListener.onClose (ForwardingClientCallListener.java:23)
    at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose (ForwardingClientCallListener.java:40)
    at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose (CensusStatsModule.java:700)
    at io.grpc.PartialForwardingClientCallListener.onClose (PartialForwardingClientCallListener.java:39)
    at io.grpc.ForwardingClientCallListener.onClose (ForwardingClientCallListener.java:23)
    at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose (ForwardingClientCallListener.java:40)
    at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose (CensusTracingModule.java:399)
    at io.grpc.internal.ClientCallImpl.closeObserver (ClientCallImpl.java:510)
    at io.grpc.internal.ClientCallImpl.access$300 (ClientCallImpl.java:66)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close (ClientCallImpl.java:630)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$700 (ClientCallImpl.java:518)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal (ClientCallImpl.java:692)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext (ClientCallImpl.java:681)
    at io.grpc.internal.ContextRunnable.run (ContextRunnable.java:37)
    at io.grpc.internal.SerializingExecutor.run (SerializingExecutor.java:123)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201 (ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run (ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    at java.lang.Thread.run (Thread.java:748)
Caused by: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 3599999079514ns. [buffered_nanos=186035342, remote_addr=spanner.googleapis.com/172.217.175.106:443]
    at io.grpc.Status.asRuntimeException (Status.java:533)
    at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose (ClientCalls.java:495)
    at io.grpc.PartialForwardingClientCallListener.onClose (PartialForwardingClientCallListener.java:39)
    at io.grpc.ForwardingClientCallListener.onClose (ForwardingClientCallListener.java:23)
    at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose (ForwardingClientCallListener.java:40)
    at com.google.cloud.spanner.spi.v1.SpannerErrorInterceptor$1$1.onClose (SpannerErrorInterceptor.java:100)
    at io.grpc.PartialForwardingClientCallListener.onClose (PartialForwardingClientCallListener.java:39)
    at io.grpc.ForwardingClientCallListener.onClose (ForwardingClientCallListener.java:23)
    at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose (ForwardingClientCallListener.java:40)
    at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose (CensusStatsModule.java:700)
    at io.grpc.PartialForwardingClientCallListener.onClose (PartialForwardingClientCallListener.java:39)
    at io.grpc.ForwardingClientCallListener.onClose (ForwardingClientCallListener.java:23)
    at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose (ForwardingClientCallListener.java:40)
    at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose (CensusTracingModule.java:399)
    at io.grpc.internal.ClientCallImpl.closeObserver (ClientCallImpl.java:510)
    at io.grpc.internal.ClientCallImpl.access$300 (ClientCallImpl.java:66)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close (ClientCallImpl.java:630)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$700 (ClientCallImpl.java:518)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal (ClientCallImpl.java:692)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext (ClientCallImpl.java:681)
    at io.grpc.internal.ContextRunnable.run (ContextRunnable.java:37)
    at io.grpc.internal.SerializingExecutor.run (SerializingExecutor.java:123)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201 (ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run (ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    at java.lang.Thread.run (Thread.java:748)`

3. My network infrastructure does not include a HTTP proxy

@olavloite Thanks for your information.
I tried the new version. But I encountered the  new error [com.google.cloud.spanner.SpannerException: INTERNAL: ] after 1 hour running.

Details information as following.
1. The stacktrace of the exception i got
 `com.google.cloud.spanner.SpannerException: INTERNAL: com.google.api.gax.rpc.InternalException: io.grpc.StatusRuntimeException: INTERNAL: Internal error encountered.
    at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerExceptionPreformatted (SpannerExceptionFactory.java:210)
    at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException (SpannerExceptionFactory.java:56)
    at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException (SpannerExceptionFactory.java:150)
    at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.get (GapicSpannerRpc.java:1214)
    at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.executePartitionedDml (GapicSpannerRpc.java:1060)
    at com.google.cloud.spanner.PartitionedDMLTransaction$1.call (PartitionedDMLTransaction.java:86)
    at com.google.cloud.spanner.PartitionedDMLTransaction$1.call (PartitionedDMLTransaction.java:68)
    at com.google.api.gax.retrying.DirectRetryingExecutor.submit (DirectRetryingExecutor.java:105)
    at com.google.cloud.RetryHelper.run (RetryHelper.java:76)
    at com.google.cloud.RetryHelper.runWithRetries (RetryHelper.java:50)
    at com.google.cloud.spanner.SpannerRetryHelper.runTxWithRetriesOnAborted (SpannerRetryHelper.java:57)
    at com.google.cloud.spanner.PartitionedDMLTransaction.executePartitionedUpdate (PartitionedDMLTransaction.java:90)
    at com.google.cloud.spanner.SessionImpl.executePartitionedUpdate (SessionImpl.java:108)
    at com.google.cloud.spanner.SessionPool$PooledSession.executePartitionedUpdate (SessionPool.java:688)
    at com.google.cloud.spanner.DatabaseClientImpl$3.apply (DatabaseClientImpl.java:205)
    at com.google.cloud.spanner.DatabaseClientImpl$3.apply (DatabaseClientImpl.java:202)
    at com.google.cloud.spanner.DatabaseClientImpl.runWithSessionRetry (DatabaseClientImpl.java:219)
    at com.google.cloud.spanner.DatabaseClientImpl.executePartitionedUpdate (DatabaseClientImpl.java:200)
    at jp.co.sej.central.spanner.DeleteOldStoreRecords.deleteUsingPartitionedDml (DeleteOldStoreRecords.java:56)
    at jp.co.sej.central.spanner.DeleteOldStoreRecords.main (DeleteOldStoreRecords.java:44)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
    at java.lang.Thread.run (Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: com.google.api.gax.rpc.InternalException: io.grpc.StatusRuntimeException: INTERNAL: Internal error encountered.
    at com.google.common.util.concurrent.AbstractFuture.getDoneValue (AbstractFuture.java:564)
    at com.google.common.util.concurrent.AbstractFuture.get (AbstractFuture.java:545)
    at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.get (GapicSpannerRpc.java:1208)
    at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.executePartitionedDml (GapicSpannerRpc.java:1060)
    at com.google.cloud.spanner.PartitionedDMLTransaction$1.call (PartitionedDMLTransaction.java:86)
    at com.google.cloud.spanner.PartitionedDMLTransaction$1.call (PartitionedDMLTransaction.java:68)
    at com.google.api.gax.retrying.DirectRetryingExecutor.submit (DirectRetryingExecutor.java:105)
    at com.google.cloud.RetryHelper.run (RetryHelper.java:76)
    at com.google.cloud.RetryHelper.runWithRetries (RetryHelper.java:50)
    at com.google.cloud.spanner.SpannerRetryHelper.runTxWithRetriesOnAborted (SpannerRetryHelper.java:57)
    at com.google.cloud.spanner.PartitionedDMLTransaction.executePartitionedUpdate (PartitionedDMLTransaction.java:90)
    at com.google.cloud.spanner.SessionImpl.executePartitionedUpdate (SessionImpl.java:108)
    at com.google.cloud.spanner.SessionPool$PooledSession.executePartitionedUpdate (SessionPool.java:688)
    at com.google.cloud.spanner.DatabaseClientImpl$3.apply (DatabaseClientImpl.java:205)
    at com.google.cloud.spanner.DatabaseClientImpl$3.apply (DatabaseClientImpl.java:202)
    at com.google.cloud.spanner.DatabaseClientImpl.runWithSessionRetry (DatabaseClientImpl.java:219)
    at com.google.cloud.spanner.DatabaseClientImpl.executePartitionedUpdate (DatabaseClientImpl.java:200)
    at jp.co.sej.central.spanner.DeleteOldStoreRecords.deleteUsingPartitionedDml (DeleteOldStoreRecords.java:56)
    at jp.co.sej.central.spanner.DeleteOldStoreRecords.main (DeleteOldStoreRecords.java:44)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
    at java.lang.Thread.run (Thread.java:748)
Caused by: com.google.api.gax.rpc.InternalException: io.grpc.StatusRuntimeException: INTERNAL: Internal error encountered.
    at com.google.api.gax.rpc.ApiExceptionFactory.createException (ApiExceptionFactory.java:67)
    at com.google.api.gax.grpc.GrpcApiExceptionFactory.create (GrpcApiExceptionFactory.java:72)
    at com.google.api.gax.grpc.GrpcApiExceptionFactory.create (GrpcApiExceptionFactory.java:60)
    at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure (GrpcExceptionCallable.java:97)
    at com.google.api.core.ApiFutures$1.onFailure (ApiFutures.java:68)
    at com.google.common.util.concurrent.Futures$CallbackListener.run (Futures.java:1050)
    at com.google.common.util.concurrent.DirectExecutor.execute (DirectExecutor.java:30)
    at com.google.common.util.concurrent.AbstractFuture.executeListener (AbstractFuture.java:1176)
    at com.google.common.util.concurrent.AbstractFuture.complete (AbstractFuture.java:969)
    at com.google.common.util.concurrent.AbstractFuture.setException (AbstractFuture.java:760)
    at io.grpc.stub.ClientCalls$GrpcFuture.setException (ClientCalls.java:526)
    at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose (ClientCalls.java:501)
    at io.grpc.PartialForwardingClientCallListener.onClose (PartialForwardingClientCallListener.java:39)
    at io.grpc.ForwardingClientCallListener.onClose (ForwardingClientCallListener.java:23)
    at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose (ForwardingClientCallListener.java:40)
    at com.google.cloud.spanner.spi.v1.SpannerErrorInterceptor$1$1.onClose (SpannerErrorInterceptor.java:100)
    at io.grpc.internal.ClientCallImpl.closeObserver (ClientCallImpl.java:426)
    at io.grpc.internal.ClientCallImpl.access$500 (ClientCallImpl.java:66)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close (ClientCallImpl.java:689)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900 (ClientCallImpl.java:577)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal (ClientCallImpl.java:751)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext (ClientCallImpl.java:740)
    at io.grpc.internal.ContextRunnable.run (ContextRunnable.java:37)
    at io.grpc.internal.SerializingExecutor.run (SerializingExecutor.java:123)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201 (ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run (ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    at java.lang.Thread.run (Thread.java:748)
Caused by: io.grpc.StatusRuntimeException: INTERNAL: Internal error encountered.
    at io.grpc.Status.asRuntimeException (Status.java:533)
    at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose (ClientCalls.java:501)
    at io.grpc.PartialForwardingClientCallListener.onClose (PartialForwardingClientCallListener.java:39)
    at io.grpc.ForwardingClientCallListener.onClose (ForwardingClientCallListener.java:23)
    at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose (ForwardingClientCallListener.java:40)
    at com.google.cloud.spanner.spi.v1.SpannerErrorInterceptor$1$1.onClose (SpannerErrorInterceptor.java:100)
    at io.grpc.internal.ClientCallImpl.closeObserver (ClientCallImpl.java:426)
    at io.grpc.internal.ClientCallImpl.access$500 (ClientCallImpl.java:66)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close (ClientCallImpl.java:689)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900 (ClientCallImpl.java:577)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal (ClientCallImpl.java:751)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext (ClientCallImpl.java:740)
    at io.grpc.internal.ContextRunnable.run (ContextRunnable.java:37)
    at io.grpc.internal.SerializingExecutor.run (SerializingExecutor.java:123)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201 (ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run (ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    at java.lang.Thread.run (Thread.java:748)
`

2. pom.xml
` <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>com.google.cloud</groupId>
        <artifactId>libraries-bom</artifactId>
        <version>5.4.0</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>

  <properties>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
  </properties>

  <dependencies>


    <!--testing dependencies-->
    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.13</version>
      <scope>test</scope>
    </dependency>

    <dependency>
      <groupId>org.hamcrest</groupId>
      <artifactId>hamcrest-library</artifactId>
      <version>2.2</version>
      <scope>test</scope>
    </dependency>

    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-databind</artifactId>
      <version>2.11.0</version>
    </dependency>

    <dependency>
      <groupId>com.google.cloud</groupId>
      <artifactId>google-cloud-spanner</artifactId>
      <version>1.55.0</version>
    </dependency>


  </dependencies>`

@olavloite I ran the job again and then the error seems different.

`com.google.cloud.spanner.AbortedException: ABORTED: com.google.api.gax.rpc.AbortedException: io.grpc.StatusRuntimeException: ABORTED: Aborted due to transient fault
    at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerExceptionPreformatted (SpannerExceptionFactory.java:196)
    at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException (SpannerExceptionFactory.java:56)
    at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException (SpannerExceptionFactory.java:150)
    at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.get (GapicSpannerRpc.java:1214)
    at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.executePartitionedDml (GapicSpannerRpc.java:1060)
    at com.google.cloud.spanner.PartitionedDMLTransaction$1.call (PartitionedDMLTransaction.java:86)
    at com.google.cloud.spanner.PartitionedDMLTransaction$1.call (PartitionedDMLTransaction.java:68)
    at com.google.api.gax.retrying.DirectRetryingExecutor.submit (DirectRetryingExecutor.java:105)
    at com.google.cloud.RetryHelper.run (RetryHelper.java:76)
    at com.google.cloud.RetryHelper.runWithRetries (RetryHelper.java:50)
    at com.google.cloud.spanner.SpannerRetryHelper.runTxWithRetriesOnAborted (SpannerRetryHelper.java:57)
    at com.google.cloud.spanner.PartitionedDMLTransaction.executePartitionedUpdate (PartitionedDMLTransaction.java:90)
    at com.google.cloud.spanner.SessionImpl.executePartitionedUpdate (SessionImpl.java:108)
    at com.google.cloud.spanner.SessionPool$PooledSession.executePartitionedUpdate (SessionPool.java:688)
    at com.google.cloud.spanner.DatabaseClientImpl$3.apply (DatabaseClientImpl.java:205)
    at com.google.cloud.spanner.DatabaseClientImpl$3.apply (DatabaseClientImpl.java:202)
    at com.google.cloud.spanner.DatabaseClientImpl.runWithSessionRetry (DatabaseClientImpl.java:219)
    at com.google.cloud.spanner.DatabaseClientImpl.executePartitionedUpdate (DatabaseClientImpl.java:200)
    at jp.co.sej.central.spanner.DeleteOldStoreRecords.deleteUsingPartitionedDml (DeleteOldStoreRecords.java:56)
    at jp.co.sej.central.spanner.DeleteOldStoreRecords.main (DeleteOldStoreRecords.java:44)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
    at java.lang.Thread.run (Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: com.google.api.gax.rpc.AbortedException: io.grpc.StatusRuntimeException: ABORTED: Aborted due to transient fault
    at com.google.common.util.concurrent.AbstractFuture.getDoneValue (AbstractFuture.java:564)
    at com.google.common.util.concurrent.AbstractFuture.get (AbstractFuture.java:545)
    at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.get (GapicSpannerRpc.java:1208)
    at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.executePartitionedDml (GapicSpannerRpc.java:1060)
    at com.google.cloud.spanner.PartitionedDMLTransaction$1.call (PartitionedDMLTransaction.java:86)
    at com.google.cloud.spanner.PartitionedDMLTransaction$1.call (PartitionedDMLTransaction.java:68)
    at com.google.api.gax.retrying.DirectRetryingExecutor.submit (DirectRetryingExecutor.java:105)
    at com.google.cloud.RetryHelper.run (RetryHelper.java:76)
    at com.google.cloud.RetryHelper.runWithRetries (RetryHelper.java:50)
    at com.google.cloud.spanner.SpannerRetryHelper.runTxWithRetriesOnAborted (SpannerRetryHelper.java:57)
    at com.google.cloud.spanner.PartitionedDMLTransaction.executePartitionedUpdate (PartitionedDMLTransaction.java:90)
    at com.google.cloud.spanner.SessionImpl.executePartitionedUpdate (SessionImpl.java:108)
    at com.google.cloud.spanner.SessionPool$PooledSession.executePartitionedUpdate (SessionPool.java:688)
    at com.google.cloud.spanner.DatabaseClientImpl$3.apply (DatabaseClientImpl.java:205)
    at com.google.cloud.spanner.DatabaseClientImpl$3.apply (DatabaseClientImpl.java:202)
    at com.google.cloud.spanner.DatabaseClientImpl.runWithSessionRetry (DatabaseClientImpl.java:219)
    at com.google.cloud.spanner.DatabaseClientImpl.executePartitionedUpdate (DatabaseClientImpl.java:200)
    at jp.co.sej.central.spanner.DeleteOldStoreRecords.deleteUsingPartitionedDml (DeleteOldStoreRecords.java:56)
    at jp.co.sej.central.spanner.DeleteOldStoreRecords.main (DeleteOldStoreRecords.java:44)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
    at java.lang.Thread.run (Thread.java:748)
Caused by: com.google.api.gax.rpc.AbortedException: io.grpc.StatusRuntimeException: ABORTED: Aborted due to transient fault
    at com.google.api.gax.rpc.ApiExceptionFactory.createException (ApiExceptionFactory.java:61)
    at com.google.api.gax.grpc.GrpcApiExceptionFactory.create (GrpcApiExceptionFactory.java:72)
    at com.google.api.gax.grpc.GrpcApiExceptionFactory.create (GrpcApiExceptionFactory.java:60)
    at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure (GrpcExceptionCallable.java:97)
    at com.google.api.core.ApiFutures$1.onFailure (ApiFutures.java:68)
    at com.google.common.util.concurrent.Futures$CallbackListener.run (Futures.java:1050)
    at com.google.common.util.concurrent.DirectExecutor.execute (DirectExecutor.java:30)
    at com.google.common.util.concurrent.AbstractFuture.executeListener (AbstractFuture.java:1176)
    at com.google.common.util.concurrent.AbstractFuture.complete (AbstractFuture.java:969)
    at com.google.common.util.concurrent.AbstractFuture.setException (AbstractFuture.java:760)
    at io.grpc.stub.ClientCalls$GrpcFuture.setException (ClientCalls.java:526)
    at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose (ClientCalls.java:501)
    at io.grpc.PartialForwardingClientCallListener.onClose (PartialForwardingClientCallListener.java:39)
    at io.grpc.ForwardingClientCallListener.onClose (ForwardingClientCallListener.java:23)
    at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose (ForwardingClientCallListener.java:40)
    at com.google.cloud.spanner.spi.v1.SpannerErrorInterceptor$1$1.onClose (SpannerErrorInterceptor.java:100)
    at io.grpc.internal.ClientCallImpl.closeObserver (ClientCallImpl.java:426)
    at io.grpc.internal.ClientCallImpl.access$500 (ClientCallImpl.java:66)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close (ClientCallImpl.java:689)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900 (ClientCallImpl.java:577)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal (ClientCallImpl.java:751)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext (ClientCallImpl.java:740)
    at io.grpc.internal.ContextRunnable.run (ContextRunnable.java:37)
    at io.grpc.internal.SerializingExecutor.run (SerializingExecutor.java:123)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201 (ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run (ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    at java.lang.Thread.run (Thread.java:748)
Caused by: io.grpc.StatusRuntimeException: ABORTED: Aborted due to transient fault
    at io.grpc.Status.asRuntimeException (Status.java:533)
    at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose (ClientCalls.java:501)
    at io.grpc.PartialForwardingClientCallListener.onClose (PartialForwardingClientCallListener.java:39)
    at io.grpc.ForwardingClientCallListener.onClose (ForwardingClientCallListener.java:23)
    at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose (ForwardingClientCallListener.java:40)
    at com.google.cloud.spanner.spi.v1.SpannerErrorInterceptor$1$1.onClose (SpannerErrorInterceptor.java:100)
    at io.grpc.internal.ClientCallImpl.closeObserver (ClientCallImpl.java:426)
    at io.grpc.internal.ClientCallImpl.access$500 (ClientCallImpl.java:66)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close (ClientCallImpl.java:689)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900 (ClientCallImpl.java:577)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal (ClientCallImpl.java:751)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext (ClientCallImpl.java:740)
    at io.grpc.internal.ContextRunnable.run (ContextRunnable.java:37)
    at io.grpc.internal.SerializingExecutor.run (SerializingExecutor.java:123)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201 (ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run (ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    at java.lang.Thread.run (Thread.java:748)
`

--- a/google-cloud-spanner/clirr-ignored-differences.xml
+++ b/google-cloud-spanner/clirr-ignored-differences.xml
@@ -170,5 +170,10 @@
     <className>com/google/cloud/spanner/spi/v1/GapicSpannerRpc</className>
     <method>com.google.spanner.v1.ResultSet executePartitionedDml(com.google.spanner.v1.ExecuteSqlRequest, java.util.Map, org.threeten.bp.Duration)</method>
   </difference>
+  <difference>
+    <differenceType>7012</differenceType>
+    <className>com/google/cloud/spanner/spi/v1/SpannerRpc</className>
+    <method>com.google.api.gax.retrying.RetrySettings getPartitionedDmlRetrySettings()</method>
+  </difference>
   
 </differences>

--- a/google-cloud-spanner/src/main/java/com/google/cloud/spanner/PartitionedDMLTransaction.java
+++ b/google-cloud-spanner/src/main/java/com/google/cloud/spanner/PartitionedDMLTransaction.java
@@ -87,7 +87,8 @@ class PartitionedDMLTransaction implements SessionTransaction {
           }
         };
     com.google.spanner.v1.ResultSet resultSet =
-        SpannerRetryHelper.runTxWithRetriesOnAborted(callable);
+        SpannerRetryHelper.runTxWithRetriesOnAborted(
+            callable, rpc.getPartitionedDmlRetrySettings());
     if (!resultSet.hasStats()) {
       throw new IllegalArgumentException(
           "Partitioned DML response missing stats possibly due to non-DML statement as input");

--- a/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerRetryHelper.java
+++ b/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerRetryHelper.java
@@ -53,6 +53,14 @@ class SpannerRetryHelper {
 
   /** Executes the {@link Callable} and retries if it fails with an {@link AbortedException}. */
   static <T> T runTxWithRetriesOnAborted(Callable<T> callable) {
+    return runTxWithRetriesOnAborted(callable, txRetrySettings);
+  }
+
+  /**
+   * Executes the {@link Callable} and retries if it fails with an {@link AbortedException} using
+   * the specific {@link RetrySettings}.
+   */
+  static <T> T runTxWithRetriesOnAborted(Callable<T> callable, RetrySettings retrySettings) {
     try {
       return RetryHelper.runWithRetries(
           callable, txRetrySettings, new TxRetryAlgorithm<>(), NanoClock.getDefaultClock());

--- a/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java
+++ b/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java
@@ -29,6 +29,7 @@ import com.google.api.gax.grpc.GrpcCallContext;
 import com.google.api.gax.grpc.InstantiatingGrpcChannelProvider;
 import com.google.api.gax.longrunning.OperationFuture;
 import com.google.api.gax.retrying.ResultRetryAlgorithm;
+import com.google.api.gax.retrying.RetrySettings;
 import com.google.api.gax.retrying.TimedAttemptSettings;
 import com.google.api.gax.rpc.AlreadyExistsException;
 import com.google.api.gax.rpc.ApiClientHeaderProvider;
@@ -217,6 +218,7 @@ public class GapicSpannerRpc implements SpannerRpc {
   private boolean rpcIsClosed;
   private final SpannerStub spannerStub;
   private final SpannerStub partitionedDmlStub;
+  private final RetrySettings partitionedDmlRetrySettings;
   private final InstanceAdminStub instanceAdminStub;
   private final DatabaseAdminStubSettings databaseAdminStubSettings;
   private final DatabaseAdminStub databaseAdminStub;
@@ -300,7 +302,7 @@ public class GapicSpannerRpc implements SpannerRpc {
 
                 // Set a keepalive time of 120 seconds to help long running
                 // commit GRPC calls succeed
-                .setKeepAliveTime(Duration.ofSeconds(GRPC_KEEPALIVE_SECONDS))
+                .setKeepAliveTime(Duration.ofSeconds(GRPC_KEEPALIVE_SECONDS * 1000))
 
                 // Then check if SpannerOptions provides an InterceptorProvider. Create a default
                 // SpannerInterceptorProvider if none is provided
@@ -336,21 +338,24 @@ public class GapicSpannerRpc implements SpannerRpc {
                   .setCredentialsProvider(credentialsProvider)
                   .setStreamWatchdogProvider(watchdogProvider)
                   .build());
+      partitionedDmlRetrySettings =
+          options
+              .getSpannerStubSettings()
+              .executeSqlSettings()
+              .getRetrySettings()
+              .toBuilder()
+              .setInitialRpcTimeout(options.getPartitionedDmlTimeout())
+              .setMaxRpcTimeout(options.getPartitionedDmlTimeout())
+              .setTotalTimeout(options.getPartitionedDmlTimeout())
+              .setRpcTimeoutMultiplier(1.0)
+              .build();
       SpannerStubSettings.Builder pdmlSettings = options.getSpannerStubSettings().toBuilder();
       pdmlSettings
           .setTransportChannelProvider(channelProvider)
           .setCredentialsProvider(credentialsProvider)
           .setStreamWatchdogProvider(watchdogProvider)
           .executeSqlSettings()
-          .setRetrySettings(
-              options
-                  .getSpannerStubSettings()
-                  .executeSqlSettings()
-                  .getRetrySettings()
-                  .toBuilder()
-                  .setInitialRpcTimeout(options.getPartitionedDmlTimeout())
-                  .setMaxRpcTimeout(options.getPartitionedDmlTimeout())
-                  .build());
+          .setRetrySettings(partitionedDmlRetrySettings);
       this.partitionedDmlStub = GrpcSpannerStub.create(pdmlSettings.build());
 
       this.instanceAdminStub =
@@ -1058,6 +1063,11 @@ public class GapicSpannerRpc implements SpannerRpc {
       ExecuteSqlRequest request, @Nullable Map<Option, ?> options) {
     GrpcCallContext context = newCallContext(options, request.getSession());
     return get(partitionedDmlStub.executeSqlCallable().futureCall(request, context));
+  }
+
+  @Override
+  public RetrySettings getPartitionedDmlRetrySettings() {
+    return partitionedDmlRetrySettings;
   }
 
   @Override

--- a/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerRpc.java
+++ b/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerRpc.java
@@ -19,6 +19,7 @@ package com.google.cloud.spanner.spi.v1;
 import com.google.api.core.ApiFuture;
 import com.google.api.core.InternalApi;
 import com.google.api.gax.longrunning.OperationFuture;
+import com.google.api.gax.retrying.RetrySettings;
 import com.google.cloud.ServiceRpc;
 import com.google.cloud.spanner.SpannerException;
 import com.google.cloud.spanner.admin.database.v1.stub.DatabaseAdminStub;
@@ -282,6 +283,8 @@ public interface SpannerRpc extends ServiceRpc {
   ResultSet executeQuery(ExecuteSqlRequest request, @Nullable Map<Option, ?> options);
 
   ResultSet executePartitionedDml(ExecuteSqlRequest request, @Nullable Map<Option, ?> options);
+
+  RetrySettings getPartitionedDmlRetrySettings();
 
   StreamingCall executeQuery(
       ExecuteSqlRequest request, ResultStreamConsumer consumer, @Nullable Map<Option, ?> options);

